controllers/vspheredeploymentzone_controller.go

```diff
@@ -264,13 +258,25 @@ func (r vsphereDeploymentZoneReconciler) reconcileDelete(ctx context.Context, de
 	})
 	if len(machinesUsingDeploymentZone) > 0 {
 		machineNamesStr := util.MachinesAsString(machinesUsingDeploymentZone.SortedByCreationTimestamp())
-		err := errors.Errorf("%s is currently in use by machines: %s", deploymentZoneCtx.VSphereDeploymentZone.Name, machineNamesStr)
-		r.Logger.Error(err, "Error deleting VSphereDeploymentZone", "name", deploymentZoneCtx.VSphereDeploymentZone.Name)
+		return errors.Errorf("error deleting VSphereDeploymentZone: %s is currently in use by machines: %s", deploymentZoneCtx.VSphereDeploymentZone.Name, machineNamesStr)
+	}
+
+	failureDomain := &infrav1.VSphereFailureDomain{}
+	failureDomainKey := client.ObjectKey{Name: deploymentZoneCtx.VSphereDeploymentZone.Spec.FailureDomain}
+	// Return an error if the FailureDomain can not be retrieved.
+	if err := r.Client.Get(ctx, failureDomainKey, failureDomain); err != nil {
+		// If the VSphereFailureDomain is not found return early and remove the finalizer.
+		// This prevents early deletion of the VSphereFailureDomain from blocking VSphereDeploymentZone deletion.
+		if apierrors.IsNotFound(err) {
+			ctrlutil.RemoveFinalizer(deploymentZoneCtx.VSphereDeploymentZone, infrav1.DeploymentZoneFinalizer)
+			return nil
+		}
```

Comment:
One thought:

Maybe we can do the following to not duplicate the finalizer removal:

```go
	if err := r.Client.Get(ctx, failureDomainKey, failureDomain); err != nil && !apierrors.IsNotFound(err) {
		return err
	}

	if failureDomain != nil {
		// updateownerref
		// delete fd
	}
```

But also ok with it as is

Reply:
:+1: Good idea

Reply:
In practice this isn't really simpler. `failureDomain` isn't `nil` at this point - it's an empty struct, so we'd have to check for something like `name == ""` or add an isEmpty() function. Aside from that the flow is about as complex IMO.

Reply:
agree, let's keep it as is then 👍